### ISSUE TITLE
Mess up torch (cuda) GPU detection if shapely.geometry imported.

### ISSUE THREAD

**Issue by ZhuYQi** [shapely/shapely#1435]
## Expected behavior and actual behavior.

```python
import torch
torch.cuda.get_device_name()
```
It'll return device name normally.
But if import shepely.geometry before import torch.
```python
from shapely.geometry import *
import torch
torch.cuda.get_device_name()
```
it returns nothing (made the process no resp), and all task on GPU will raise a segment fault when executing `venv/lib/python3.7/site-packages/torch/cuda/__init__.py:216`
```python
torch._C._cuda_init()
```

## Steps to reproduce the problem.

`python3 -c "from shapely.geometry import *; import torch; torch.cuda.get_device_name()"`

## Operating system

Ubuntu2004 5.4.0-121-generic 
Nvidia Driver Version: 510.39.01    
CUDA Version: 11.6 
Python 3.7.13 (default, Apr 24 2022, 01:05:22) [GCC 9.4.0] on linux

## Shapely version and provenance

Shapely 1.8.2       installed from PyPI using pip
torch     1.11.0     installed from PyPI using pip

**Comment by ZhuYQi** [comment]
maybe shapely.speedups.enable() change some configuration of cython.

**Comment by rskew** [comment]
There seems to be an issue with version 1.8.2 on PyPI: I can reliably reproduce this issue after `pip install shapely==1.8.2`, however the issue goes away when installing from the github repo:
`pip3 uninstall shapely; pip3 install git+https://github.com/shapely/shapely.git@1.8.2`

Workarounds I've found now include:
- installing shapely from github
- importing torch before importing shapely

Interestingly, reproducing the bug while probing with strace and gdb show that `torch._c._cuda_init()` is attempting to read from file descriptor 0 (stdin) when using std::random_device to generate a random seed. This read from stdin (rather than `/dev/urandom` or `/dev/random` as it should) causes execution to block.

**Comment by jorisvandenbossche** [comment]
> I can reliably reproduce this issue after pip install shapely==1.8.2, however the issue goes away when installing from the github repo:

The main difference between both methods is that if you install from PyPI, you (normally) get a wheel that has GEOS bundled (currently version 3.10.2). While if you install from source from github, you are building shapely against a GEOS library that is available on your system (and which is apparently installed, since the build worked). 

(I don't know how that necessarily explains why the one triggers the issue with torch and the other not, but so that is the main difference I can think of)

Could you check what version of GEOS you have installed on your system? (`geos-config --version` on the command line)


---

What would also be interesting is if you could check if `pip install --pre shapely==2.0a1` still gives the same problem or not (the version 2 is a complete refactor of the internals, and eg no longer uses ctypes at runtime)

**Comment by rskew** [comment]
Hi @jorisvandenbossche ,

Running `geos-config --version` gives `3.8.0`

Interesting that intalling from github builds against the available system GEOS, although the interaction with pytorch is still mysterious.

Running `pip install --pre shapely==2.0a1` indeed fixes the problem, I'll run with that as a solution.

Thanks for your help :) let me know if there's anything else you want me to try, or if you want more complete steps to reproduce the problem.

**Comment by ZhuYQi** [comment]
> > I can reliably reproduce this issue after pip install shapely==1.8.2, however the issue goes away when installing from the github repo:
> 
> The main difference between both methods is that if you install from PyPI, you (normally) get a wheel that has GEOS bundled (currently version 3.10.2). While if you install from source from github, you are building shapely against a GEOS library that is available on your system (and which is apparently installed, since the build worked).
> 
> (I don't know how that necessarily explains why the one triggers the issue with torch and the other not, but so that is the main difference I can think of)
> 
> Could you check what version of GEOS you have installed on your system? (`geos-config --version` on the command line)
> 
> What would also be interesting is if you could check if `pip install --pre shapely==2.0a1` still gives the same problem or not (the version 2 is a complete refactor of the internals, and eg no longer uses ctypes at runtime)

**Comment by ZhuYQi** [comment]
> maybe shapely.speedups.enable() changed some configuration of cython.

https://github.com/shapely/shapely/issues/1598#issuecomment-1311678011
